### PR TITLE
Tree widget: Add ContextMenu to trees

### PR DESCRIPTION
**Pull request by saskliutas**
Closes https://github.com/iTwin/viewer-components-react/issues/494

### CODE REVIEW

File: packages/itwin/tree-widget/src/components/trees/models-tree/ModelsTree.tsx [hunk 1]
```diff
@@ -38,10 +39,16 @@ export interface ModelsTreeHierarchyConfiguration {
 }
 
 /**
- * Props for [[ModelsTree]] component
+ * Base props for [[ModelsTree]] component.
  * @public
  */
-export interface ModelsTreeProps extends BaseFilterableTreeProps {
+export type ModelsTreeBaseProps = BaseFilterableTreeProps & TreeContextMenuProps;
```
Comment (grigasp) [inline packages/itwin/tree-widget/src/components/trees/models-tree/ModelsTree.tsx:45]: do we need  / use this type anywhere besides using it below?

File: packages/itwin/tree-widget/src/test/trees/TreeRenderer.test.tsx [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+      />
+    );
+
+    const nodeElement = await waitFor(() => getByText("Test Node"));
+    await user.pointer({ keys: "[MouseRight>]", target: nodeElement });
+
+    await waitFor(() => getByText("Test Item"));
+  });
+
+  it("doesn't open context menu if there are no items", async () => {
```
Comment (grigasp) [inline packages/itwin/tree-widget/src/test/trees/TreeRenderer.test.tsx:54]: please add a similar test for the case when `contextMenuItems=[() => null]`

**Review by grigasp — CHANGES_REQUESTED** [review]

**Comment by grigasp** [comment]
Just to confirm... I suppose you didn't add context menu support to `IModelContentTree` and `ExternalSourcesTree` because they don't have the same UX and `ModelsTree` and `CategoryTree` (don't use the same node renderers)?

**Comment by saskliutas** [comment]
> Just to confirm... I suppose you didn't add context menu support to `IModelContentTree` and `ExternalSourcesTree` because they don't have the same UX and `ModelsTree` and `CategoryTree` (don't use the same node renderers)?

No, I think I just missed those trees. I made it possible to use ContextMenu without all the visibility tree styling so added context menu to other trees too.

File: packages/itwin/tree-widget/src/test/trees/TreeRenderer.test.tsx (hunk 2 above)
Comment (saskliutas) [reply packages/itwin/tree-widget/src/test/trees/TreeRenderer.test.tsx:54]: Added

File: packages/itwin/tree-widget/src/components/trees/models-tree/ModelsTree.tsx (hunk 1 above)
Comment (saskliutas) [reply packages/itwin/tree-widget/src/components/trees/models-tree/ModelsTree.tsx:45]: No it used as base type for `ModelsTreeProps`.

File: packages/itwin/tree-widget/src/components/trees/VisibilityTreeRenderer.tsx [hunk 3]
```diff
@@ -23,10 +25,18 @@ import type { VisibilityTreeFilterInfo } from "./Common";
...
+export interface VisibilityTreeRendererProps extends TreeRendererBaseProps {
+  nodeRendererProps: VisibilityTreeNodeRendererProps;
+}
+
+/**
+ * Props for base tree node renderer used by visibility trees.
  * @public
  */
-export interface VisibilityTreeRendererProps {
+export interface VisibilityTreeNodeRendererProps {
```
[the remark this answers is not in the document]
Comment (saskliutas) [reply packages/itwin/tree-widget/src/components/trees/VisibilityTreeRenderer.tsx]: Done in https://github.com/iTwin/viewer-components-react/pull/554

**Comment by aruniverse** [comment]
Would it be possible to update so that the entry point for the context menu is shown on hover? I noticed for the property-grid yesterday, only shows when you right click

File: packages/itwin/tree-widget/src/components/trees/models-tree/ModelsTree.tsx (hunk 1 above)
Comment (saskliutas) [reply packages/itwin/tree-widget/src/components/trees/models-tree/ModelsTree.tsx]: Applied suggestion

**Comment by grigasp** [comment]
> Would it be possible to update so that the entry point for the context menu is shown on hover? I noticed for the property-grid yesterday, only shows when you right click

Can you give more details on why you think that's needed? Or give some example where it's implemented? I looked at several components in multiple applications and don't see this feature anywhere. Or maybe I just misunderstand what you're asking...

**Comment by aruniverse** [comment]
> Can you give more details on why you think that's needed? Or give some example where it's implemented? I looked at several components in multiple applications and don't see this feature anywhere. Or maybe I just misunderstand what you're asking...

I haven't seen any examples of this in our apps either, might not have looked enough.

I found the "entry point" for our context menus a little hard to find. I noticed in google drive the other day they always show an ellipses for an entry point to their context menu. And on hover they show actions you can take like favorite and etc.

image

Here's the property grid right now with just the copy properties:
image

idk how users will know they will be able to copy the cell value unless they accidentally right click on the row

If you look at teams when you hover over a contacts profile and then proceed to hover over some information it'll show an icon to copy which i think is more intuitive 
image

I dont necessarily think we need to extract the copy to outside the context menu, just think the entry point (rather than right clicking) showing an ellipses on hover would be better UX

**Comment by grigasp** [comment]
Thanks. I believe we should consult with UX first, so let's complete this PR and we can follow up with discoverability improvement afterwards (filed a separate [issue](https://github.com/iTwin/viewer-components-react/issues/556))

**Review by grigasp — APPROVED** [review]
